Add more to caching docs and make them top level (#14290)
PrefectHQ/prefect: docs/3.0rc/develop/task-caching.mdx

    print('running...')
    return x + 42

my_cached_task(1)
my_cached_task(1, debug=True) # still uses the cache
```

### Cache key functions

You can configure custom cache policy logic through the use of cache key functions.
A cache key function is a function that accepts two positional arguments:
- The first argument corresponds to the `TaskRunContext`, which stores task run metadata. For example, 
this object has attributes `task_run_id`, `flow_run_id`, and `task`, all of which can be used in your
custom logic.
- The second argument corresponds to a dictionary of input values to the task. For example, 
if your task has the signature `fn(x, y, z)` then the dictionary will have keys "x", "y", and "z" with corresponding values that can be used to compute your cache key.

This function can then be specified using the `cache_key_fn` argument on 
the [task decorator](https://prefect-python-sdk-docs.netlify.app/prefect/tasks/#prefect.tasks.task).

For example:

```python
def static_cache_key(context, parameters):
    # return a constant
    return "static cache key"

@task(cache_key_fn=static_cache_key)
def my_cached_task(x: int):
    return x + 1
```

## Multi-task caching

> There are many situations in which multiple tasks need to always run together as an atomic unit.

Reviewer: ```suggestion
There are many situations in which multiple tasks need to always run together or not at all.
```

> There are many situations in which multiple tasks need to always run together as an atomic unit.

Author: That works for me

```python
@task(cache_policy=TASK_SOURCE)
def my_stateful_task():
    print('sleeping')
    time.sleep(10)

    # change the return value, for example
    return 43 

my_stateful_task() # sleeps again
```

## Customizing the cache

Prefect allows users to configure task caching behavior in numerous ways.

### Cache expiration

All cache keys can optionally be given an _expiration_ through the `cache_expiration` keyword on 
the [task decorator](https://prefect-python-sdk-docs.netlify.app/prefect/tasks/#prefect.tasks.task):
This keyword accepts a `datetime.timedelta` specifying a duration for which the cached value should be
considered valid.

Providing an expiration value results in Prefect persisting an expiration timestamp alongside the result
record for the task.
This expiration is then applied to _all_ other tasks that may share this cache key.

### Cache policies

Cache policies can be composed and altered using basic Python syntax to form more complex policies.
For example, all task policies except for `NONE` can be _added_ together to form new policies that combine
the individual policies' logic into a larger cache key computation.

> Combining policies in this way therefore results in caches that are _easier_ to bust.

Reviewer 2: ```suggestion
Combining policies in this way results in caches that are _easier_ to invalidate.
```

For example:

```python
from prefect import task
from prefect.cache_policies import TASK_SOURCE, INPUTS

> 

Reviewer 2: ```suggestion

```

@task(cache_policy=TASK_SOURCE + INPUTS)
def my_cached_task(x: int):
    return x + 42
```

This task will rerun anytime you provide new values for `x`, _or_ anytime you change the underlying code.

> The `INPUTS` policy is a special policy that allows you to _subtract_ string values to signal ignoring

Reviewer 2: ```suggestion
The `INPUTS` policy is a special policy that allows you to _subtract_ string values to ignore
```

    print('running...')
    return x + 42

my_cached_task(1)
my_cached_task(1, debug=True) # still uses the cache
```

### Cache key functions

Users can configure custom cache policy logic through the use of cache key functions.
A cache key function is a function that accepts two positional arguments:
- The first argument corresponds to the `TaskRunContext`, which stores task run metadata. For example, 
this object has attributes `task_run_id`, `flow_run_id`, and `task`, all of which can be used in your
custom logic.
- The second argument corresponds to a dictionary of input values to the task. For example, 
if your task has the signature `fn(x, y, z)` then the dictionary will have keys "x", "y", and "z" with corresponding values that can be used to compute your cache key.

This function can then be specified using the `cache_key_fn` argument on 
the [task decorator](https://prefect-python-sdk-docs.netlify.app/prefect/tasks/#prefect.tasks.task).

For example:

```python
def static_cache_key(context, parameters):
    # return a constant
    return "static cache key"

@task(cache_key_fn=static_cache_key)
def my_cached_task(x: int):
    return x + 1
```

> ## Multi-task caching

Author: This is the only new section - the rest was cut/pasted directly from the write tasks doc

There are many situations in which multiple tasks need to always run together as an atomic unit.
This can be achieved in Prefect by configuring these tasks to always write to their caches within 
a single [_transaction_](/3.0rc/develop/transactions).

```python
from prefect import task, flow
from prefect.transactions import transaction

@task(cache_key_fn=lambda *args, **kwargs: "static-key-1")
def load_data():
    return "some-data"

@task(cache_key_fn=lambda *args, **kwargs: "static-key-2")
def process_data(data, fail):
    if fail:
        raise RuntimeError("Error! Abort!")

    return len(data)

@flow
def multi_task_cache(fail: bool = True):
    with transaction():
        data = load_data()
        process_data(data=data, fail=fail)
```

When this flow is run with default parameter values it will fail on the `process_data` task.
The `load_data` task will succeed. However, because caches are only written to when a transaction
is _committed_, the `load_data` task will _not_ write a result to its cache key location until
the `process_data` task succeeds as well.

This ensures that anytime you need to rerun this flow both `load_data` and `process_data` are executed
together. 

> And after a successful execution both will be indefinitely cached until the cache key is updated.

Reviewer 2: ```suggestion
After a successful execution both tasks will be cached until the cache key is updated.
```

> For more information on transactions, see [here](/3.0rc/develop/transactions). 

Reviewer 2: ```suggestion
Read more about [transactions](/3.0rc/develop/transactions). 
```

## Caching example

> In this example, until the `cache_expiration` time ends, as long as the input to `hello_task()` remains 

Reviewer 2: ```suggestion
In this example, until the `cache_expiration` time is reached, as long as the input to `hello_task()` remains 
```

> the same when it is called, the cached return value is returned. In this situation the task is not rerun. 

Reviewer 2: ```suggestion
the same when it is called, the cached return value will be returned. The task is not rerun. 
```

However, if the input argument value changes, `hello_task()` runs using the new input.

```python 
from datetime import timedelta
from prefect import flow, task
from prefect.cache_policies import INPUTS

> 

Reviewer 2: ```suggestion

```
